### ISSUE TITLE
GUnicorn/Flask server is not running with AzureLogHandler

### ISSUE THREAD

**Issue by freedev** [census-instrumentation/opencensus-python#957]
**Describe your environment.** 

We want to export Python logs to Azure Monitor. 
When AzureLogHandler runs into gunicorn, only main process use the handler, others log but not nothing reaches azure. 
In other words when you make a request to flask, methods were running into child processes (threads or forks) and such methods were unable to log through the handler. 

Current environment is running into a docker container with:
- ubuntu:18.04
- Python 3.6.9 
- Flask==1.1.2
- gunicorn==20.0.4
- opencensus-ext-azure==1.0.4
- opencensus-ext-flask==0.7.3

**Steps to reproduce.**
I have created a minimal project on github where reproduce the problem:

    git clone https://github.com/freedev/flask-azure-application-insights
    cd flask-azure-application-insights
    ./run-docker.sh 00000000-0000-0000-0000-000000000000

modify 00000000-0000-0000-0000-000000000000 with your own InstrumentationKey 
Once the docker instance is started there is first log message written by main process.
Each requests that comes to flask application should be logged through the AzureLogHandler but none arrives to Application Insights

try with:

    curl http://localhost:5000/logging

**What is the expected behavior?**
We expect to be able to use AzureLogHandler and write into Application Insights

**What is the actual behavior?**
Each requests that comes to flask application should be logged through the AzureLogHandler but none arrives to Application Insights

**Additional context.**

**Comment by lzchen** [comment]
See [this comment](https://github.com/census-instrumentation/opencensus-python/issues/928#issuecomment-671489488) for something that is related. It seems that gunicorn uses [os.fork](https://github.com/benoitc/gunicorn/issues/1570#issuecomment-323419791) to spawn it's child processes, so the worker thread (in your parent process) that is responsible for exporting your telemetry is not copied.

You can either try having the initialization logic (`AzureLogHandler`) be initialized newly in all of your child processes (put the logic in where the fork will happen), or try to use the `multiprocessing` module like [here](https://github.com/census-instrumentation/opencensus-python/issues/928#issuecomment-679497439).

**Comment by freedev** [comment]
Hi @lzchen, thanks for your suggestion. 
I suppose that the change of context would be a problem and sharing the Azure Log Handler between different processes would be a problem. 
So I tried to configure GUnicorn using multiple thread instead of process, but without any luck. Please see the `GU_THREADS_NUM` and `GU_WORK_NUM` env variables defined into the `entrypoint.sh` script.

But I even tried to create a new logger (and a new AzureLogHandler) for each process/thread but even this solutions seems to fail (here I've added a branch named [instance_per_pid](https://github.com/freedev/flask-azure-application-insights/tree/instance_per_pid) where I try to craete a new logger for each pid).

**Comment by lzchen** [comment]
I believe `GU_WORK_NUM ` represents the number of workers that are spun up (each which are a new process). In run-docker.sh, what values are you using for `GU_THREADS_NUM` and `GU_WORK_NUM`?

Also, can you specify exactly what you do when you are testing creating a new logger instance per process/thread and what you are seeing in your App insights telemetry to indicate that it is not working?

**Comment by freedev** [comment]
I've tried different configuration once using more threads and leaving `GU_WORK_NUM` at 1 or using more processes and leaving `GU_THREADS_NUM` at 1

let's say I've tried `GU_THREADS_NUM=3` and `GU_WORK_NUM=1` or `GU_THREADS_NUM=1` and `GU_WORK_NUM=3`

In any case, the logger works correctly, i.e. I see in stdout the lines logged by the app. It's only the AzureLogHandler that does only works in the main thread/process. 

Is there somewhere the source code of the AzureLogHandler in order to see what's happening inside?

**Comment by freedev** [comment]
Thanks @lzchen, trying to create a new logger for each pid (process or thread) has worked.
Now I have an instance AzureLogHandler working for each process. 

Instead of having a global `logger` now I've created a `getLogger()` function that returns for each process/thread its own logger, the working implementation is here [master branch](https://github.com/freedev/flask-azure-application-insights/blob/master/application.py#L34)

**Comment by freedev** [comment]
> I believe `GU_WORK_NUM ` represents the number of workers that are spun up (each which are a new process). 

yep

> In run-docker.sh, what values are you using for `GU_THREADS_NUM` and `GU_WORK_NUM`?

I've tried with GU_THREADS_NUM=3 and GU_WORK_NUM=1 and GU_WORK_NUM=3 and GU_THREADS_NUM=1
As said, having a new logger instance (and a new AzureLogHandler) for each pid has figured out the problem.

`getLogger()` works well with more workers (GU_WORK_NUM > 1), I see a new line in the log analytics for each process id. 
On the other hand, when I try with multiple threads, the load I'm generating with my browser it's not enough to generate many threads. I think I've to write a test unit to double check this.

**Comment by freedev** [comment]
BTW, reading at row generated in the log analytics I see the AzureLogHandler try to track where is running, I mean in which process. Have a look at this screenshot.

<img width="999" alt="Screenshot 2020-10-06 at 11 44 46" src="https://user-images.githubusercontent.com/464254/95185765-6d63a580-07c9-11eb-9930-9a72a98869f7.png">

**Comment by lzchen** [comment]
> BTW, reading at row generated in the log analytics I see the AzureLogHandler try to track where is running, I mean in which process. Have a look at this screenshot. The customDimensions field is generated automatically by the log handler.

Oh that is useful information. Appreciate the help.

I'm glad you got it working! Is this issue resolved?

**Comment by freedev** [comment]
Hi @lzchen, not sure if I can consider the issue resolved. The mileage may vary :)
Just to understand if we can accept this solution, I'm trying to imagine what will happen at scale.
Consider there are many microservices and using this approach, each microservice has a number logger depending from how is the concurrency configured. 
How many AzureLogHandler can I have concurrently opened? 
Each of them open a connection? 
How many connection the Log Analyzer server handles? Should I open an new issue trying to understand this?

**Comment by lzchen** [comment]
@freedev 
Yes those concerns make perfect sense. The way the Azure exporters are designed today, I do not believe that there is a different workaround that would not generate equivalent concerns. Might I suggest opening a feature request to explore possible multi-processing solutions in OpenCensus?

**Comment by freedev** [comment]
> Yes those concerns make perfect sense. The way the Azure exporters are designed today, I do not believe that there is a different workaround that would not generate equivalent concerns. Might I suggest opening a feature request to explore possible multi-processing solutions in OpenCensus?

Yep, I will.

> Just an FYI: Take a look at [this link](https://github.com/census-instrumentation/opencensus-python/issues/928#issuecomment-679497439). It seemed like using a multiprocessing queue in the exporter worked. Could you possibly test your code using this?

Again, yes. I remember this link comes from the comment you gave me in first instance and now, with hindsight, it seems very helpful.

**Comment by lzchen** [comment]
That's great. Are you able to mark the ticket submitted to Azure devops as resolved? If this is a new feature, we can continue the conversation here or in the new feature request.

**Comment by br0cky** [comment]
Hi both, stumbled on this as i am facing the same issue. However i noticed removing the --preload flag on my gunicorn command seems to allow logging to App Insights. @freedev have you seen this also?

**Comment by omsharma2364** [comment]
@freedev I was facing similar issue. Got it working by using multiprocessing-logging with minimal changes. Usage as mentioned in [https://pypi.org/project/multiprocessing-logging/]. Just make sure you call install_mp_handler() after you have added all the desired handlers.

**Comment by jonasmiederer** [comment]
I faced the same problem and solved it by [replacing the queue in the scheduler with a shared managed queue](https://github.com/jonasmiederer/opencensus-python/blob/master/opencensus/common/schedule/__init__.py#L86). Everything seems to work fine so far.
